python/cudf/cudf/core/frame.py

```diff
@@ -1750,6 +1750,106 @@ def __arrow_array__(self, type=None):
 
+    def round(self, decimals=0):
+        """
+        Round a DataFrame to a variable number of decimal places.
+
+        Parameters
+        ----------
+        decimals : int, dict, Series
+            Number of decimal places to round each column to. If an int is
+            given, round each column to the same number of places.
+            Otherwise dict and Series round to variable numbers of places.
+            Column names should be in the keys if `decimals` is a
+            dict-like, or in the index if `decimals` is a Series. Any
+            columns not included in `decimals` will be left as is. Elements
+            of `decimals` which are not columns of the input will be
+            ignored.
+        deep : boolean, optional, default False
+            Whether to make deep copy or shallow copy of the columns.
+
+        Returns
+        -------
+        DataFrame
+            A DataFrame with the affected columns rounded to the specified
+            number of decimal places.
+
+        Examples
+        --------
+        >>> df = cudf.DataFrame(
+                [(.21, .32), (.01, .67), (.66, .03), (.21, .18)],
+        ...     columns=['dogs', 'cats']
+        ... )
+        >>> df
+            dogs  cats
+        0  0.21  0.32
+        1  0.01  0.67
+        2  0.66  0.03
+        3  0.21  0.18
+
+        By providing an integer each column is rounded to the same number
+        of decimal places
+
+        >>> df.round(1)
+            dogs  cats
+        0   0.2   0.3
+        1   0.0   0.7
+        2   0.7   0.0
+        3   0.2   0.2
+
+        With a dict, the number of places for specific columns can be
+        specified with the column names as key and the number of decimal
+        places as value
+
+        >>> df.round({'dogs': 1, 'cats': 0})
+            dogs  cats
+        0   0.2   0.0
+        1   0.0   1.0
+        2   0.7   0.0
+        3   0.2   0.0
+
+        Using a Series, the number of places for specific columns can be
+        specified with the column names as index and the number of
+        decimal places as value
+
+        >>> decimals = cudf.Series([0, 1], index=['cats', 'dogs'])
+        >>> df.round(decimals)
+            dogs  cats
+        0   0.2   0.0
+        1   0.0   1.0
+        2   0.7   0.0
+        3   0.2   0.0
+        """
+
+        copy_data = self._data.copy(deep=True)
+
+        if isinstance(decimals, cudf.Series):
+            decimals = decimals.to_pandas()
+
+        if isinstance(decimals, (dict, pd.Series)):
+            if (
+                isinstance(decimals, pd.Series)
+                and not decimals.index.is_unique
+            ):
+                raise ValueError("Index of decimals must be unique")
+            for name, col in copy_data.items():
+                if (
+                    pd.api.types.is_numeric_dtype(col.dtype)
+                    and name in decimals.keys()
+                ):
+                    copy_data[name] = col.round(decimals[name])
```

Comment:
Correct me if I'm wrong.. I think this implementation for arg type `Series` keeps all values at host. And every call of `col.round` incurs a H2D? Would recommend coercing to cudf dataframe instead of pandas dataframe:

```suggestion
        if isinstance(decimals, pd.Series):
            decimals = cudf.from_pandas(decimals)

        if isinstance(decimals, (dict, cudf.Series)):
            if (
                isinstance(decimals, cudf.Series)
                and not decimals.index.is_unique
            ):
                raise ValueError("Index of decimals must be unique")
            for name, col in copy_data.items():
                if (
                    pd.api.types.is_numeric_dtype(col.dtype)
                    and name in decimals.keys()
                ):
                    copy_data[name] = col.round(decimals[name])
```

CC @rgsl888prabhu @shwina

Reply:
`decimals[name]` will do D->H and again inside libcudf H->D if we are using cudf Scalar for every element, in other case you will get pandas series at once from D->H, which I feel reduces a lots of burden.

In future if we return `cudf.Scalar` when requesting scalar values from a series then your approach would be an awesome solution.

Reply:
From offline discussion: indexing from cudf domain currently has more overhead than doing it in Pandas. We also assumed that `Series` type argument should be relatively small in size (the same order of magnitude of `dict`). Thus performance implication will be low to do it in pandas.